File: cpp_merge/classified/packages/plugin-todo/src/actions/completeTodo.cpp
#include "completeTodo.hpp"

std::shared_ptr<Promise<std::shared_ptr<TaskCompletion>>> extractTaskCompletion(std::shared_ptr<IAgentRuntime> runtime, std::shared_ptr<Memory> message, array<std::shared_ptr<TodoData>> availableTasks, std::shared_ptr<State> state)
{
    try
    {
        auto tasksText = availableTasks->map([=](auto task) mutable
        {
            return std:("ID: ") + task->id + std:("\
Name: ") + task->name + std:("\
Description: ") + (OR((task->description), (task->name))) + std:("\
Tags: ") + (OR((task->tags->join(std:(", "))), (std:("none")))) + std:("\
");
        }
        )->join(std:("\
---\
"));
        auto messageHistory = formatMessages(object{
            object::pair{std:("messages"), OR((state->data->messages), (array<any>()))}, 
            object::pair{std:("entities"), OR((state->data->entities), (array<any>()))}
        });
        auto prompt = composePrompt(object{
            object::pair{std:("state"), object{
                object::pair{std:("text"), OR((message->content->text), (string_empty))}, 
                object::pair{std:("availableTasks"), tasksText}, 
                object::pair{std:("messageHistory"), std:("messageHistory")}
            }}, 
            object::pair{std:("template"), extractCompletionTemplate}
        });
        auto result = std::async([=]() { runtime->useModel(ModelType->TEXT_SMALL, object{
            object::pair{std:("prompt"), std:("prompt")}, 
            object::pair{std:("stopSequences"), array<any>()}
        }); });
        auto parsedResult = as<any>(parseKeyValueXml(result));
        if (OR((!parsedResult), (type_of(parsedResult->isFound) == std:("undefined")))) {
            logger->error(std:("Failed to parse valid task completion information from XML"));
            return object{
                object::pair{std:("taskId"), string_empty}, 
                object::pair{std:("taskName"), string_empty}, 
                object::pair{std:("isFound"), false}
            };
        }
        auto finalResult = object{
            object::pair{std:("taskId"), (parsedResult->taskId == std:("null")) ? any(string_empty) (String(OR((parsedResult->taskId), (string_empty))))}, 
            object::pair{std:("taskName"), (parsedResult->taskName == std:("null")) ? any(string_empty) (String(OR((parsedResult->taskName), (string_empty))))}, 
            object::pair{std:("isFound"), String(parsedResult->isFound) == std:("true")}
        };
        return finalResult;
    }
    catch (const any& error)
    {
        logger->error(std:("Error extracting task completion information:"), error);
        return object{
            object::pair{std:("taskId"), string_empty}, 
            object::pair{std:("taskName"), string_empty}, 
            object::pair{std:("isFound"), false}
        };
    }
};


string extractCompletionTemplate = std:("\
# Task: Extract Task Completion Information\
\
## User Message\
{{text}}\
\
## Message History\
{{messageHistory}}\
\
## Available Tasks\
{{availableTasks}}\
\
## Instructions\
Parse the user's message to identify which task they're marking as completed.\
Match against the list of available tasks by name or description.\
If multiple tasks have similar names, choose the closest match.\
\
Return an XML object with:\
<response>\
  <taskId>ID of the task being completed, or 'null' if not found</taskId>\
  <taskName>Name of the task being completed, or 'null' if not found</taskName>\
  <isFound>'true' or 'false' indicating if a matching task was found</isFound>\
</response>\
\
## Example Output Format\
<response>\
  <taskId>123e4567-e89b-12d3-a456-426614174000</taskId>\
  <taskName>Finish report</taskName>\
  <isFound>true</isFound>\
</response>\
\
If no matching task was found:\
<response>\
  <taskId>null</taskId>\
  <taskName>null</taskName>\
  <isFound>false</isFound>\
</response>\
");
std::shared_ptr<Action> completeTodoAction = object{
    object::pair{std:("name"), std:("COMPLETE_TODO")}, 
    object::pair{std:("similes"), array<string>{ std:("FINISH_TODO"), std:("DONE_TODO"), std:("MARK_TODO_DONE"), std:("TODO_COMPLETE") }}, 
    object::pair{std:("description"), std:("Mark a todo item as completed, including tasks done in the real world. Works with task name or ID. Can be chained with LIST_TODOS to show remaining tasks.")}, 
    object::pair{std:("validate"), [=](auto _runtime, auto _message) mutable
    {
        return true;
    }
    }, 
    object::pair{std:("handler"), [=](auto runtime, auto message, auto state, auto options, auto callback) mutable
    {
        logger->info(std:("[completeTodo] Handler invoked"), object{
            object::pair{std:("messageContent"), message->content->text}
        });
        auto dataService = as<std::shared_ptr<TodoService>>(runtime->getService(std:("todo")));
        if (!dataService) {
            logger->error(std:("[completeTodo] Todo service not available"));
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), std:("Sorry, the todo service is not available right now. Please try again later.")}, 
                    object::pair{std:("error"), true}, 
                    object::pair{std:("source"), message->content->source}
                }); });
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("values"), object{
                    object::pair{std:("error"), std:("Todo service not available")}
                }}
            };
        }
        try
        {
            auto availableTasks = std::async([=]() { dataService->getTodos(object{
                object::pair{std:("entityId"), message->entityId}, 
                object::pair{std:("roomId"), message->roomId}, 
                object::pair{std:("isCompleted"), false}
            }); });
            if (availableTasks->get_length() == 0) {
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), std:("You don't have any incomplete tasks to mark as done. Would you like to create a new task?")}, 
                        object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO_NO_TASKS") }}, 
                        object::pair{std:("source"), message->content->source}
                    }); });
                }
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("actionName"), std:("COMPLETE_TODO")}, 
                        object::pair{std:("error"), std:("No incomplete tasks found")}
                    }}, 
                    object::pair{std:("values"), object{
                        object::pair{std:("success"), false}, 
                        object::pair{std:("hasActiveTasks"), false}
                    }}
                };
            }
            shared taskCompletion = (options["taskId"]) ? any(object{
                object::pair{std:("taskId"), options["taskId"]}, 
                object::pair{std:("taskName"), options["taskName"]}, 
                object::pair{std:("isFound"), true}
            }) (std::async([=]() { extractTaskCompletion(runtime, message, availableTasks, state); }));
            if (!taskCompletion["isFound"]) {
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), std:("I couldn't determine which task you're marking as completed. Could you be more specific? Here are your current tasks:\
\
") + availableTasks->map([=](auto task) mutable
                        {
                            return std:("- ") + task->name + string_empty;
                        }
                        )->join(std:("\
")) + string_empty}, 
                        object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO_NOT_FOUND") }}, 
                        object::pair{std:("source"), message->content->source}
                    }); });
                }
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("actionName"), std:("COMPLETE_TODO")}, 
                        object::pair{std:("error"), std:("Could not identify which task to complete")}
                    }}, 
                    object::pair{std:("values"), object{
                        object::pair{std:("success"), false}, 
                        object::pair{std:("availableTaskCount"), availableTasks->get_length()}
                    }}
                };
            }
            auto task = availableTasks->find([=](auto t) mutable
            {
                return t->id == taskCompletion["taskId"];
            }
            );
            if (!task) {
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), std:("I couldn't find a task matching "") + taskCompletion["taskName"] + std:("". Please try again with the exact task name.")}, 
                        object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO_NOT_FOUND") }}, 
                        object::pair{std:("source"), message->content->source}
                    }); });
                }
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("actionName"), std:("COMPLETE_TODO")}, 
                        object::pair{std:("error"), std:("Task not found: ") + taskCompletion["taskName"] + string_empty}, 
                        object::pair{std:("searchedTaskId"), taskCompletion["taskId"]}
                    }}, 
                    object::pair{std:("values"), object{
                        object::pair{std:("success"), false}
                    }}
                };
            }
            std::async([=]() { dataService->updateTodo(task->id, object{
                object::pair{std:("isCompleted"), true}, 
                object::pair{std:("completedAt"), std::make_shared<Date>()}, 
                object::pair{std:("metadata"), utils::assign(object{
                    , 
                    object::pair{std:("completedAt"), ((std::make_shared<Date>()))->toISOString()}
                }, task->metadata)}
            }); });
            auto responseText = string_empty;
            if (task->type == std:("daily")) {
                responseText = std:("✅ Daily task completed: "") + task->name + std:(""");
            } else if (task->type == std:("one-off")) {
                auto completedOnTime = (task->dueDate) ? any(std::make_shared<Date>() <= task->dueDate) (true);
                auto timeStatus = (completedOnTime) ? std:("on time") : std:("late");
                auto priority = OR((task->priority), (4));
                responseText = std:("✅ Task completed: "") + task->name + std:("" (Priority ") + priority + std:(", ") + timeStatus + std:(")");
            } else if (task->type == std:("aspirational")) {
                responseText = std:("🌟 Congratulations on achieving your aspirational goal: "") + task->name + std:(""!\
\
This is a significant accomplishment.");
            } else {
                responseText = std:("✅ Marked "") + task->name + std:("" as completed.");
            }
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), responseText}, 
                    object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO") }}, 
                    object::pair{std:("source"), message->content->source}
                }); });
            }
            return object{
                object::pair{std:("success"), true}, 
                object::pair{std:("data"), object{
                    object::pair{std:("actionName"), std:("COMPLETE_TODO")}, 
                    object::pair{std:("taskId"), task->id}, 
                    object::pair{std:("taskName"), task->name}, 
                    object::pair{std:("taskType"), task->type}, 
                    object::pair{std:("completedAt"), ((std::make_shared<Date>()))->toISOString()}
                }}, 
                object::pair{std:("values"), object{
                    object::pair{std:("success"), true}, 
                    object::pair{std:("taskId"), task->id}, 
                    object::pair{std:("taskName"), task->name}, 
                    object::pair{std:("taskType"), task->type}
                }}
            };
        }
        catch (const any& error)
        {
            logger->error(std:("Error in completeTodo handler:"), error);
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), std:("I encountered an error while completing your task. Please try again.")}, 
                    object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO_ERROR") }}, 
                    object::pair{std:("source"), message->content->source}
                }); });
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("data"), object{
                    object::pair{std:("actionName"), std:("COMPLETE_TODO")}, 
                    object::pair{std:("error"), (is<Error>(error)) ? any(error->message) (std:("Unknown error"))}
                }}, 
                object::pair{std:("values"), object{
                    object::pair{std:("success"), false}
                }}
            };
        }
    }
    }, 
    object::pair{std:("examples"), as<array<array<std::shared_ptr<ActionExample>>>>(array<array<std::shared_ptr<ActionExample>>>{ array<object>{ object{
        object::pair{std:("name"), std:("{{name1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I completed my taxes")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{name2}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("✅ Task completed: "Finish taxes" (Priority 2, on time)")}, 
            object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO") }}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("{{name1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I did my 50 pushups today")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{name2}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("✅ Daily task completed: "Do 50 pushups"")}, 
            object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO") }}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("{{name1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I read three books this month")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{name2}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("🌟 Congratulations on achieving your aspirational goal: "Read more books"!\
\
This is a significant accomplishment.")}, 
            object::pair{std:("actions"), array<string>{ std:("COMPLETE_TODO") }}
        }}
    } } })}
};

void Main(void)
{
}

MAIN

File: classified/packages/plugin-todo/src/actions/completeTodo.ts
import {
  type Action,
  type ActionExample,
  composePrompt,
  formatMessages,
  HandlerCallback,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  parseKeyValueXml,
  type State,
} from '@elizaos/core';
import type { TodoData, TodoService } from '../services/todoService';

// Interface for task completion properties
interface TaskCompletion {
  taskId: string;
  taskName: string;
  isFound: boolean;
}

/**
 * Template for extracting task completion information from user message
 */
const extractCompletionTemplate = `
# Task: Extract Task Completion Information

## User Message
{{text}}

## Message History
{{messageHistory}}

## Available Tasks
{{availableTasks}}

## Instructions
Parse the user\'s message to identify which task they\'re marking as completed.\nMatch against the list of available tasks by name or description.\nIf multiple tasks have similar names, choose the closest match.\n\nReturn an XML object with:\n<response>\n  <taskId>ID of the task being completed, or \'null\' if not found</taskId>\n  <taskName>Name of the task being completed, or \'null\' if not found</taskName>\n  <isFound>\'true\' or \'false\' indicating if a matching task was found</isFound>\n</response>\n\n## Example Output Format
<response>\n  <taskId>123e4567-e89b-12d3-a456-426614174000</taskId>\n  <taskName>Finish report</taskName>\n  <isFound>true</isFound>\n</response>\n\nIf no matching task was found:\n<response>\n  <taskId>null</taskId>\n  <taskName>null</taskName>\n  <isFound>false</isFound>\n</response>\n`;

/**
 * Extracts which task the user wants to mark as completed
 */
async function extractTaskCompletion(
  runtime: IAgentRuntime,
  message: Memory,
  availableTasks: TodoData[],
  state: State
): Promise<TaskCompletion> {
  try {
    // Format available tasks for the prompt
    const tasksText = availableTasks
      .map((task) => {
        return `ID: ${task.id}\nName: ${task.name}\nDescription: ${task.description || task.name}\nTags: ${task.tags?.join(', ') || 'none'}\n`;
      })
      .join('\n---\n');

    const messageHistory = formatMessages({
      messages: state.data?.messages || [],
      entities: state.data?.entities || [],
    });

    const prompt = composePrompt({
      state: {
        text: message.content.text || '',
        availableTasks: tasksText,
        messageHistory,
      },
      template: extractCompletionTemplate,
    });

    const result = await runtime.useModel(ModelType.TEXT_SMALL, {
      prompt,
      stopSequences: [],
    });

    // Parse XML from the text results
    const parsedResult = parseKeyValueXml(result) as TaskCompletion | null;

    if (!parsedResult || typeof parsedResult.isFound === 'undefined') {
      logger.error('Failed to parse valid task completion information from XML');
      return { taskId: '', taskName: '', isFound: false };
    }

    // Convert string 'true'/'false' to boolean and handle 'null' strings
    const finalResult: TaskCompletion = {
      taskId: parsedResult.taskId === 'null' ? '' : String(parsedResult.taskId || ''),
      taskName: parsedResult.taskName === 'null' ? '' : String(parsedResult.taskName || ''),
      isFound: String(parsedResult.isFound) === 'true',
    };

    return finalResult;
  } catch (error) {
    logger.error('Error extracting task completion information:', error);
    return { taskId: '', taskName: '', isFound: false };
  }
}

/**
 * The COMPLETE_TODO action allows users to mark a task as completed.
 */
export const completeTodoAction: Action = {
  name: 'COMPLETE_TODO',
  similes: ['FINISH_TODO', 'DONE_TODO', 'MARK_TODO_DONE', 'TODO_COMPLETE'],
  description:
    'Mark a todo item as completed, including tasks done in the real world. Works with task name or ID. Can be chained with LIST_TODOS to show remaining tasks.',

  validate: async (_runtime: IAgentRuntime, _message: Memory): Promise<boolean> => {
    return true;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    options: any,
    callback: HandlerCallback | undefined
  ) => {
    logger.info('[completeTodo] Handler invoked', { messageContent: message.content.text });

    const dataService = runtime.getService('todo') as TodoService;
    if (!dataService) {
      logger.error('[completeTodo] Todo service not available');
      if (callback) {
        await callback({
          text: 'Sorry, the todo service is not available right now. Please try again later.',
          error: true,
          source: message.content.source,
        });
      }
      return {
        success: false,
        values: {
          error: 'Todo service not available',
        },
      };
    }

    try {
      // Get available tasks
      const availableTasks = await dataService.getTodos({
        entityId: message.entityId,
        roomId: message.roomId,
        isCompleted: false,
      });

      if (availableTasks.length === 0) {
        if (callback) {
          await callback({
            text: "You don't have any incomplete tasks to mark as done. Would you like to create a new task?",
            actions: ['COMPLETE_TODO_NO_TASKS'],
            source: message.content.source,
          });
        }
        return {
          success: false,
          data: {
            actionName: 'COMPLETE_TODO',
            error: 'No incomplete tasks found',
          },
          values: {
            success: false,
            hasActiveTasks: false,
          },
        };
      }

      // Extract which task the user wants to complete
      const taskCompletion = options?.taskId
        ? { taskId: options.taskId, taskName: options.taskName, isFound: true }
        : await extractTaskCompletion(runtime, message, availableTasks, state!);

      if (!taskCompletion.isFound) {
        if (callback) {
          await callback({
            text: `I couldn't determine which task you're marking as completed. Could you be more specific? Here are your current tasks:\n\n${availableTasks
              .map((task) => `- ${task.name}`)
              .join('\n')}`,
            actions: ['COMPLETE_TODO_NOT_FOUND'],
            source: message.content.source,
          });
        }
        return {
          success: false,
          data: {
            actionName: 'COMPLETE_TODO',
            error: 'Could not identify which task to complete',
          },
          values: {
            success: false,
            availableTaskCount: availableTasks.length,
          },
        };
      }

      // Find the task in the available tasks
      const task = availableTasks.find((t) => t.id === taskCompletion.taskId);

      if (!task) {
        if (callback) {
          await callback({
            text: `I couldn't find a task matching "${taskCompletion.taskName}". Please try again with the exact task name.`,
            actions: ['COMPLETE_TODO_NOT_FOUND'],
            source: message.content.source,
          });
        }
        return {
          success: false,
          data: {
            actionName: 'COMPLETE_TODO',
            error: `Task not found: ${taskCompletion.taskName}`,
            searchedTaskId: taskCompletion.taskId,
          },
          values: {
            success: false,
          },
        };
      }

      // Mark the task as completed
      await dataService.updateTodo(task.id, {
        isCompleted: true,
        completedAt: new Date(),
        metadata: {
          ...task.metadata,
          completedAt: new Date().toISOString(),
        },
      });

      // Generate response text based on task type
      let responseText = '';

      if (task.type === 'daily') {
        responseText = `✅ Daily task completed: "${task.name}"`;
      } else if (task.type === 'one-off') {
        const completedOnTime = task.dueDate ? new Date() <= task.dueDate : true;
        const timeStatus = completedOnTime ? 'on time' : 'late';
        const priority = task.priority || 4;

        responseText = `✅ Task completed: "${task.name}" (Priority ${priority}, ${timeStatus})`;
      } else if (task.type === 'aspirational') {
        responseText = `🌟 Congratulations on achieving your aspirational goal: "${task.name}"!\n\nThis is a significant accomplishment.`;
      } else {
        responseText = `✅ Marked "${task.name}" as completed.`;
      }

      if (callback) {
        await callback({
          text: responseText,
          actions: ['COMPLETE_TODO'],
          source: message.content.source,
        });
      }

      return {
        success: true,
        data: {
          actionName: 'COMPLETE_TODO',
          taskId: task.id,
          taskName: task.name,
          taskType: task.type,
          completedAt: new Date().toISOString(),
        },
        values: {
          success: true,
          taskId: task.id,
          taskName: task.name,
          taskType: task.type,
        },
      };
    } catch (error) {
      logger.error('Error in completeTodo handler:', error);
      if (callback) {
        await callback({
          text: 'I encountered an error while completing your task. Please try again.',
          actions: ['COMPLETE_TODO_ERROR'],
          source: message.content.source,
        });
      }
      return {
        success: false,
        data: {
          actionName: 'COMPLETE_TODO',
          error: error instanceof Error ? error.message : 'Unknown error',
        },
        values: {
          success: false,
        },
      };
    }
  },

  examples: [
    [
      {
        name: '{{name1}}',
        content: {
          text: 'I completed my taxes',
        },
      },
      {
        name: '{{name2}}',
        content: {
          text: '✅ Task completed: "Finish taxes" (Priority 2, on time)',
          actions: ['COMPLETE_TODO'],
        },
      },
    ],
    [
      {
        name: '{{name1}}',
        content: {
          text: 'I did my 50 pushups today',
        },
      },
      {
        name: '{{name2}}',
        content: {
          text: '✅ Daily task completed: "Do 50 pushups"',
          actions: ['COMPLETE_TODO'],
        },
      },
    ],
    [
      {
        name: '{{name1}}',
        content: {
          text: 'I read three books this month',
        },
      },
      {
        name: '{{name2}}',
        content: {
          text: '🌟 Congratulations on achieving your aspirational goal: "Read more books"!\n\nThis is a significant accomplishment.',
          actions: ['COMPLETE_TODO'],
        },
      },
    ],
  ] as ActionExample[][],
};

export default completeTodoAction;
